### ISSUE TITLE
[jit] Proposal: fuse add ops that expand tensors

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#10096]
Let's say we have `torch.add(self, other)` and `self` has size `(64, 2048)`, while other has size `(2048)`. This isn't fusable into a FusionGroup because the inputs (self, other) have different sizes.

However, `other` can be expanded into the size of `self`. If we insert an `expand` node, then the operation will be fusable into a FusionGroup. The cost of an expand is around 3 microseconds, which is less than launching a kernel specifically for adding `self, other` (launch overhead of 6-12 microseconds).

This matters for LSTM perf: the following graph is the nvprof output for a lstm cell's forward pass.
![image](https://user-images.githubusercontent.com/5652049/43489620-9b9a95b0-94eb-11e8-850f-f4199c9d51d5.png)

The two kernels after the second sgemm are add kernels that "should" be fusable into the FusionGroup kernel on the right.

cc @zdevito @apaszke

**Comment by ngimel** [comment]
That's strange, as of a few days back the following function
```
@torch.jit.trace(torch.rand(4,4),  torch.rand(4), torch.rand(4))
def scaleshift(x, scale, shift):
    return x * scale + shift
```
where `scale` and `shift` are broadcastable to x was properly fused into a single kernel for me. Important difference is that in my case `scale` and `shift` did not require gradient, which is not the case for adding bias in LSTM. For a function to be put in a fusion block it has to be symbolically differentiable, IIRC, and expand is not.

**Comment by zou3519** [comment]
There used to be explicit expands before relevant binary operations. Something changed somewhere (my guess is the TensorIterator got rid of the necessity for the explicit expand?) and now there are no explicit expands:
```
In [1]: import torch

In [2]: @torch.jit.trace(torch.rand(4,4),  torch.rand(4), torch.rand(4))
   ...: def scaleshift(x, scale, shift):
   ...:     return x * scale + shift
   ...:
   ...:

In [3]: scaleshift.graph
Out[3]:
graph(%0 : Float(4, 4)
      %1 : Float(4)
      %2 : Float(4)) {
  %3 : Float(4, 4) = aten::mul(%0, %1)
  %4 : int = prim::Constant[value=1]()
  %5 : Float(4, 4) = aten::add(%3, %2, %4)
  return (%5);
}

In [4]: x = torch.randn(4, 4)

In [5]: y = torch.randn(4)

In [6]: z = torch.randn(4)

In [7]: scaleshift(x, y, z)
Out[7]:
tensor([[-5.4599, -0.7210,  0.1930,  1.1432],
        [-0.5724,  3.6090,  0.0220,  1.2254],
        [ 1.6685,  2.4105,  0.5106,  1.0230],
        [-2.2619,  3.7919,  0.1854,  1.2218]])

In [8]: scaleshift.graph_for(x, y, z)
Out[8]:
graph(%0 : Float(4, 4)
      %1 : Float(4)
      %2 : Float(4)) {
  %3 : Float(4, 4) = aten::mul(%0, %1)
  %4 : int = prim::Constant[value=1]()
  %5 : Float(4, 4) = aten::add(%3, %2, %4)
  return (%5);
}
```

Without explicit expands, the inputs don't have the same sizes anymore so no fusion occurs.

**Comment by ngimel** [comment]
You are right, couple weeks back tracer inserted explicit expands, now it does not.

**Comment by soumith** [comment]
yes, this is a sweeping change that @colesbury made. now operators are responsible for broadcasting, I think.

**Comment by zou3519** [comment]
What changed is that we don't have explicit expands anymore; some of the operators (add, mul, sub, and div, and possibly others?) can do the broadcasting themselves.

I thought about two potential solutions for the graph fuser to adapt to this:
1) We could have a pass to add explicit expands before the graph fuser pass and a pass to remove explicit expands after. This way all fusion opportunities are available to the graph fuser
2) The graph fuser can emit expand nodes if it would otherwise be able to fuse a simple mappable operation aside from the detail that the inputs are different sizes

**Comment by apaszke** [comment]
Me, @zdevito and @colesbury discussed that when merging the `TensorIterator` PR, and decided to settle for a temporary regression in the fuser to unblock it. I don't think we should insert unnecessary expands if the nodes won't get fused anyway, because that's just adding extra work, so I'd go with option 2.
